### PR TITLE
Ias xsuaa token exchange

### PR DESCRIPTION
**Pull request by liga-oz**

### CODE REVIEW

File: spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerResolverTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+		request = new MockHttpServletRequest();
+		tokenBroker = new TokenBrokerResolver(getXsuaaServiceConfiguration(), null,
+				oAuth2TokenService,
+				authenticationConfiguration);
+	}
+
+	@Test
+	public void testXsuaaTokenCache() throws OAuth2ServiceException {
+		oAuth2TokenService = spy(oAuth2TokenService);
+		oAuth2TokenService.retrieveAccessTokenViaJwtBearerTokenGrant(any(), any(), any(), any(), any());
```
Comment (nenaraab) [inline spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerResolverTest.java:51]: does this influence the counter?

File: spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerResolverTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+		XsuaaServiceConfigurationDummy cfg = new XsuaaServiceConfigurationDummy();
+		cfg.appId = "a1!123";
+		cfg.clientId = "myclient!t1";
+		cfg.clientSecret = "top.secret";
+		cfg.uaaDomain = "auth.com";
+		cfg.uaaUrl = "https://mydomain.auth.com";
+		return cfg;
+	}
+
+	public AuthenticationInformationExtractor authenticationMethods(AuthenticationMethod... methods) {
```
Comment (nenaraab) [inline spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerResolverTest.java:86]: required?

File: spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerResolverTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import com.sap.cloud.security.xsuaa.XsuaaServiceConfiguration;
+import com.sap.cloud.security.xsuaa.client.OAuth2ServiceException;
+import com.sap.cloud.security.xsuaa.client.OAuth2TokenService;
+
+@RunWith(SpringRunner.class)
+@ContextConfiguration(classes = { XsuaaServiceConfigurationDummy.class,
```
Comment (nenaraab) [inline spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerResolverTest.java:26]: do you have to refer to it?

File: spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerResolverTest.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+		TokenBrokerTestConfiguration.class })
```
Comment (nenaraab) [inline spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerResolverTest.java:27]: you do not pass cache into TokenBrokerResolver

File: spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerTestConfiguration.java [hunk 5]
```diff
@@ -92,7 +93,7 @@ public OAuth2TokenResponse retrieveAccessTokenViaJwtBearerTokenGrant(URI tokenEn
 					ClientCredentials clientCredentials, String token, @Nullable String subdomain,
 					@Nullable Map<String, String> optionalParameters, boolean disableCacheForRequest)
 					throws OAuth2ServiceException {
-				return null;
+				return new OAuth2TokenResponse(XSUAA_TOKEN, 100, null);
```
Comment (nenaraab) [inline spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerTestConfiguration.java:96]: required?

File: spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/extractor/IasXsuaaExchangeBroker.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	 * @param encodedJwtToken Encoded token to be checked
+	 * @return true if provided token is a XSUAA token
+	 */
+	public boolean isXsuaaToken(String encodedJwtToken) {
+		String claims = Base64JwtDecoder.getInstance().decode(encodedJwtToken).getPayload();
+		try {
+			JSONObject externalAttributeClaim = new JSONObject(claims)
+					.getJSONObject(com.sap.cloud.security.token.TokenClaims.XSUAA.EXTERNAL_ATTRIBUTE);
+			String externalAttributeValue = externalAttributeClaim
+					.getString(com.sap.cloud.security.token.TokenClaims.XSUAA.EXTERNAL_ATTRIBUTE_ENHANCER);
```
Comment (nenaraab) [inline spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/extractor/IasXsuaaExchangeBroker.java:42]: static import

File: spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/extractor/IasXsuaaExchangeBroker.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	 *
+	 * @param iasToken encoded IAS token
+	 * @return encoded Xsuaa token
+	 */
+	@Nullable
+	public String getXsuaaToken(String iasToken) {
+		try {
+			return xsuaaTokenFlows.userTokenFlow().token(iasToken).execute().getAccessToken();
+		} catch (TokenFlowException e) {
+			e.printStackTrace();
```
Comment (nenaraab) [inline spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/extractor/IasXsuaaExchangeBroker.java:60]: pls improve, add logging stmt.

File: spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/extractor/IasXsuaaExchangeBroker.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+		}
+		return null;
+	}
+
+	/**
+	 * Checks environment variable 'XSUAA_IAS_XCHANGE_ENABLED' if token exchange
+	 * between IAS and XSUAA is enabled. If 'XSUAA_IAS_XCHANGE_ENABLED' is set to 'false'
+	 * token exchange is disregarded. Any other value except null is interpreted as true.
```
Comment (nenaraab) [inline spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/extractor/IasXsuaaExchangeBroker.java:68]: If XSUAA_IAS_XCHANGE_ENABLED is not provided or with an empty value or with value = false then its disregarded.

File: spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerResolver.java [hunk 9]
```diff
@@ -33,6 +34,9 @@
  * {@code X-Identity-Zone-Subdomain} must be set (or the
  * AuthenticationInformationExtractor needs to be implemented).
  *
+ * Token exchange between IAS and XSUAA is enabled by default.
```
Comment (nenaraab) [inline spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerResolver.java:37]: is not enabled...

**Review by nenaraab — CHANGES_REQUESTED** [review]
Hi @liga-oz 
thanks a lot for this PR.
I only have some few and minor review comments.
Best regards,
Nena

File: spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerTestConfiguration.java (hunk 5 above)
Comment (liga-oz) [reply spring-xsuaa-it/src/test/java/com/sap/cloud/security/xsuaa/extractor/TokenBrokerTestConfiguration.java:96]: required for TokenBrokerResolverTest when IAS token is provided

**Review by nenaraab — APPROVED** [review]

**Comment by nenaraab** [comment]
perfect, you're done :-)

**Pull request merged**